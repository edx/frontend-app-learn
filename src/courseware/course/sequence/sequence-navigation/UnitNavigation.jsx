import classNames from 'classnames';
import PropTypes from 'prop-types';
import { useIntl } from '@edx/frontend-platform/i18n';

import { GetCourseExitNavigation } from '../../course-exit';

import { useSequenceNavigationMetadata } from './hooks';
import messages from './messages';
import PreviousButton from './generic/PreviousButton';
import NextButton from './generic/NextButton';
import { NextUnitTopNavTriggerSlot } from '../../../../plugin-slots/NextUnitTopNavTriggerSlot';

const UnitNavigation = ({
  sequenceId,
  unitId,
  onClickPrevious,
  onClickNext,
  isAtTop,
  courseId,
}) => {
  const intl = useIntl();
  const {
    isFirstUnit, isLastUnit, nextLink, previousLink,
  } = useSequenceNavigationMetadata(sequenceId, unitId);

  const renderPreviousButton = () => (
    <PreviousButton
      isFirstUnit={isFirstUnit}
      variant="outline-secondary"
      buttonLabel={intl.formatMessage(messages.previousButton)}
      buttonStyle="previous-button justify-content-center"
      onClick={onClickPrevious}
      previousLink={previousLink}
    />
  );

  const renderNextButton = () => {
    const { exitActive, exitText } = GetCourseExitNavigation(courseId, intl);
    const buttonText = (isLastUnit && exitText) ? exitText : intl.formatMessage(messages.nextButton);
    const disabled = isLastUnit && !exitActive;
    const variant = 'outline-primary';
    const buttonStyle = 'next-button justify-content-center';

    if (isAtTop) {
      return (
        <NextUnitTopNavTriggerSlot
          {...{
            courseId,
            variant,
            buttonStyle,
            buttonText,
            disabled,
            sequenceId,
            nextLink,
            onClickHandler: onClickNext,
          }}
        />
      );
    }

    return (
      <NextButton
        variant={variant}
        buttonStyle={buttonStyle}
        onClickHandler={onClickNext}
        disabled={disabled}
        buttonText={buttonText}
        nextLink={nextLink}
        hasEffortEstimate
      />
    );
  };

  return (
    <div className={classNames('unit-navigation d-flex', { 'top-unit-navigation mb-3 w-100': isAtTop })}>
      {renderPreviousButton()}
      {renderNextButton()}
    </div>
  );
};

UnitNavigation.propTypes = {
  courseId: PropTypes.string.isRequired,
  sequenceId: PropTypes.string.isRequired,
  unitId: PropTypes.string,
  onClickPrevious: PropTypes.func.isRequired,
  onClickNext: PropTypes.func.isRequired,
  isAtTop: PropTypes.bool,
};

UnitNavigation.defaultProps = {
  unitId: null,
  isAtTop: false,
};

export default UnitNavigation;
