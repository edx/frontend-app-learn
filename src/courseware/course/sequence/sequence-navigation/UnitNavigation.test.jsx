import React from 'react';
import { fireEvent } from '@testing-library/dom';
import {
  initialState, render, screen, testUnits,
} from '../../../../setupTest';
import UnitNavigation from './UnitNavigation';

describe('Unit Navigation', () => {
  const mockData = {
    sequenceId: '1',
    unitId: '2',
    onClickPrevious: () => {},
    onClickNext: () => {},
  };

  it('renders correctly without units', () => {
    const { asFragment } = render(<UnitNavigation
      {...mockData}
      sequenceId=""
      unitId=""
      onClickPrevious={() => {}}
      onClickNext={() => {}}
    />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles the clicks', () => {
    const onClickPrevious = jest.fn();
    const onClickNext = jest.fn();

    render(<UnitNavigation
      {...mockData}
      sequenceId=""
      unitId=""
      onClickPrevious={onClickPrevious}
      onClickNext={onClickNext}
    />);

    fireEvent.click(screen.getByRole('button', { name: /previous/i }));
    expect(onClickPrevious).toHaveBeenCalledTimes(1);

    fireEvent.click(screen.getByRole('button', { name: /next/i }));
    expect(onClickNext).toHaveBeenCalledTimes(1);
  });

  it('should have the navigation buttons enabled for the non-corner unit in the sequence', () => {
    render(<UnitNavigation {...mockData} />, { initialState });
    screen.getAllByRole('button').forEach(button => {
      expect(button).toBeEnabled();
    });
  });

  it('should have the "Previous" button disabled for the first unit in the sequence', () => {
    render(<UnitNavigation {...mockData} unitId="1" />, { initialState });
    expect(screen.getByRole('button', { name: /previous/i })).toBeDisabled();
    expect(screen.getByRole('button', { name: /next/i })).toBeEnabled();
  });

  it('should display "learn.end.of.course" message instead of the "Next" button for the last unit in the sequence', () => {
    render(
      <UnitNavigation
        {...mockData}
        sequenceId="2"
        unitId={testUnits.length.toString()}
      />, { initialState },
    );
    expect(screen.getByRole('button', { name: /previous/i })).toBeEnabled();
    expect(screen.queryByRole('button', { name: /next/i })).not.toBeInTheDocument();
    expect(screen.getByText("You've reached the end of this course!")).toBeInTheDocument();
  });
});
