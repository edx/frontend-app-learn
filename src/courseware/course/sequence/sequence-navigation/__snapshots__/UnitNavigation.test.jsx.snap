// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Navigation renders correctly without units 1`] = `
<DocumentFragment>
  <div
    class="unit-navigation d-flex"
  >
    <button
      class="btn btn-outline-secondary previous-button mr-2"
      type="button"
    >
      <img
        alt="fa-chevron-left"
        class="fa-chevron-left"
        data-testid="icon"
      />
      <span>
        Previous
      </span>
    </button>
    <button
      class="btn btn-outline-primary next-button"
      type="button"
    >
      <span>
        Next
      </span>
      <img
        alt="fa-chevron-right"
        class="fa-chevron-right"
        data-testid="icon"
      />
    </button>
  </div>
</DocumentFragment>
`;
